// @useall/usealltestframework/js/commumLocators.js
module.exports = {
    geral: {
        campoPesquisar: 'input[data-test-id="filtro_pesquisa_geral"]',
        campoModeloRelX: "//input[@data-test-id='ModeloSelecionado.Identificacao']",
        campoModeloRelatorioTriggerX: "//input[@data-test-id= 'ModeloSelecionado.Identificacao']/parent::div/following-sibling::div",
        btnPesquisar: 'span[data-test-id="btn_pesquisar"]',
        btnNovo: "span[data-test-id='btn_novo']",
        btnSalvar: "span[data-test-id='btn_salvar']",
        btnSalvarX: "//span[@data-test-id='btn_salvar']",
        btnCancelar: "span[data-test-id='btn_cancelar']",
        btnCancelarX: "//span[@data-test-id='btn_cancelar']",
        btnMaisOpcoesX: "//span[contains(text(),'Mais opções')]",
        btnFechar: "span[id^='use-closebutton'][id$='btnInnerEl']",
        btnVisualizar: "span[id^='use-splitbutton'][id$='btnInnerEl']",
        btnVisualizarArrow: "span[id^='use-splitbutton'][id$='arrowEl']",
        btnVisualizarArrowVisualizarExcelX: "//span[text()='Visualizar com Excel']",
        btnListas: "div[id^='tile-geral-listas'][id$='innerCt']",
        cabecalhoColunaCodigoX: "//div[starts-with(@id, 'gridcolumn')]//span[text()= 'Código']",
        menuRncArrowX: "//div[starts-with(@id, 'mainWest-innerCt')] //div[@data-qtip='Não conformidades']/preceding-sibling::div[@class='x-mi mi-arrow-drop-down']",
        listaSearchfield: "[id$='-picker-listWrap'] ul li",
        listaSearchfieldVazia: "[id$='-picker-listWrap'] ul",
        menuAtivoX: "//div[@class='item-menu active ']",
        toast: ".x-toast",
        loadmask: "[aria-describedby^='loadmask']",
        loadmaskRelX: "//div[text()= 'Emitindo relatório']",
        messageBox: "div[id^='messagebox'][id$='msg'], div[id^='use-messagebox'][id$='msg']",
        btnSimMessageBoxX: "//div[starts-with(@id, 'use-messagebox')] //span[contains(text(),'Sim')]",
        btnNaoMessageBoxX: "//div[starts-with(@id, 'use-messagebox')] //span[contains(text(),'Não')]",
        linkAtivo: "label[id^='use-linkativo']",
        fotoUsuario: "img[alt='Foto usuário']",
        janelaF2: "div[id*='pesquisageral-panel'][style*='block']",
        janelaF2Fechada: "div[id*='pesquisageral-panel'][style*='none']",
        campoBuscaF2: "div[id*='pesquisageral-panel'] input",
        expandirFiltroRelatorio: "div[class^='x-tool'][id^='fieldset'][id$='legendToggle']",
        iframeRel: "iframe[id='pdf']",
        bodyLoader: "body[class*='aguardando']",
    },

    documentos: {
        listagem: "div[id^= 'lista-documentos']",
        btnPesquisar: "span[id^='use-searchbutton-'][id$='-btnIconEl']",
        campoPesquisar: "input[data-test-id='DocsFilter.PesquisaGeral']",
        colunaIcon: "div[id^='lista-documentos'] .x-grid-item:nth-child(1) .x-grid-cell:nth-child(1)",
        colunaNome: "div[id^='lista-documentos'] .x-grid-item:nth-child(1) .x-grid-cell:nth-child(3)",
    },

    login: {
        campoUsuario: "input[id='usuario']",
        campoSenha: "input[id='senha']",
        campoBanco: "select[id='banco']",
        btnLogar: "input[id='submitFormLogin']",
    },

    mensageria: {
        iconeMensagens: "div.mi-email",
        listaMensagens: "div[id^='listatemplate-geral-mensagens-'][id$='-innerCt']",
        painelNovaMensagem: "div[id^='panel-base-novamensagem']",
        campoParaItens: "ul[id^='tagfield-usuariosemails'] .x-boundlist-item",
        dropDownCampoPara: ".x-boundlist ul[role='listbox']",
        loadmaskX: "//body/div/div[@class=css-shadow]",
        janelaMensageria: "div[id^='panel-mensagens']",
        campoPara: "input[data-test-id='Destinatarios']",
        tagCampoPara: ".x-tagfield-item",
        tagCampoParaFechar: ".x-tagfield-item-close",
        campoParaCopia: "input[data-test-id='DadosDestinatariosCc.Codigos']",
        campoParaCopiaOculta: "input[data-test-id='DadosDestinatariosCco.Codigos']",
        campoAssunto: "input[data-test-id='Mensagem.Assunto']",
        checkPrioridadeAlta: "input[data-test-id='Mensagem.PrioridadeAlta']",
        checkConfirmarRecebimento: "input[data-test-id='Mensagem.ConfirmarRecebimento']",
        btnEnviar: "span[data-test-id='btn_salvar']",
        btnCancelar: "span[data-test-id='btn_cancelar']",
        btnInserirAnexo: "input[id^='use-multifilefield-'][id$='-button-fileInputEl']",
        iframeMensagem: "iframe[id^='useall-htmleditor']",
        iframeTinymce: "body[id='tinymce'] div:nth-child(1)",
        btnExpandirDestinatarios: "div[id^='use-icon-'][id$='-innerCt']",
        btnLupaDestinatarios: "div[id^='tagfield-usuariosemails-'][id$='-trigger-pesquisar']",
        novaMensagemButtom: "div[id^='panel-mensagens'] a[id^='use-button']",
        btnFecharLupa: "div[id^='window-base-buscausuariosemails-'] .x-tool-close",
        gridLupa: "div[id^='lista-base-buscausuariosemails-usuarios'] .x-grid-item-container",
        campoNomeLupa: "div[id^='lista-base-buscausuariosemails-usuarios'] .x-grid-item-container",
        btnPesquisarLupa: "div[id^='window-base-buscausuariosemails-'] span[id^='use-searchbutton-'][id$='-btnInnerEl']",
        valCheckLupa: "div[id^='lista-base-buscausuariosemails-usuarios'][id$='body'] .x-grid-item:nth-child(1) .x-grid-cell:nth-child(1)",
        valCodigoLupa: "div[id^='lista-base-buscausuariosemails-usuarios'][id$='body'] .x-grid-item:nth-child(1) .x-grid-cell:nth-child(2)",
        valNomeLupa: "div[id^='lista-base-buscausuariosemails-usuarios'][id$='body'] .x-grid-item:nth-child(1) .x-grid-cell:nth-child(3)",
        valTipoLupa: "div[id^='lista-base-buscausuariosemails-usuarios'][id$='body'] .x-grid-item:nth-child(1) .x-grid-cell:nth-child(4)",
        btnSelecionarLupa: "div[id^='window-base-buscausuariosemails-'] span[id^='use-selectbutton-'][id$='-btnInnerEl']",
        btnCancelarLupa: "div[id^='window-base-buscausuariosemails-'] span[id^='use-cancelbutton-'][id$='-btnInnerEl']",
        menuEntradaX: "//span[starts-with(text(), 'Entrada')]",
        menuEnviadasX: "//span[starts-with(text(), 'Enviadas')]",
        menuArquivadasX: "//span[starts-with(text(), 'Arquivadas')]",
        menuSpamX: "//span[starts-with(text(), 'Spam')]",
        flagNenhumaMsg: "div[id^='listatemplate-geral-mensagens'] [class='x-mi mi-drafts']",
        mensagemX: "//div[starts-with(@id, 'listatemplate-geral-mensagens')]//div[@data-action='onClickVisualizarMsg']",
        linhaPrimeiraMsgX: "//div[starts-with(@id, 'listatemplate-geral-mensagens')]//div[@data-action='onClickVisualizarMsg'][1]",
        colFotoPrimeiraMsgX: "//div[starts-with(@id, 'listatemplate-geral-mensagens')]//div[@data-action='onClickVisualizarMsg'][1]/div[1]",
        fotoX: "//img[@class= 'abreviaturaEmail']",
        checkedX: "//div[@class='x-mi mi-check-box font-color-lightblue']",
        uncheckedX: "//div[@class='x-mi mi-check-box-outline-blank']",
        colNomePrimeiraMsgX: "//div[starts-with(@id, 'listatemplate-geral-mensagens')]//div[@data-action='onClickVisualizarMsg'][1]/div[2]",
        colAssuntoPrimeiraMsgX: "//div[starts-with(@id, 'listatemplate-geral-mensagens')]//div[@data-action='onClickVisualizarMsg'][1]/div[3]",
        colHoraPrimeiraMsgX: "//div[starts-with(@id, 'listatemplate-geral-mensagens')]//div[@data-action='onClickVisualizarMsg'][1]/div[4]",
        arquivarLinhaPrimeiraMsgX: "//div[starts-with(@id, 'listatemplate-geral-mensagens')]//div[@data-qtip='Arquivar']",
        desarquivarLinhaPrimeiraMsgX: "//div[starts-with(@id, 'listatemplate-geral-mensagens')]//div[@data-qtip='Desarquivar']",
        marcarLidoLinhaPrimeiraMsgX: "//div[starts-with(@id, 'listatemplate-geral-mensagens')]//div[@data-qtip='Marcar como lido/não lido']",
        arquivarTopoX: "//div[starts-with(@id, 'tbar-lista-mensagens')]//a[@data-qtip='Arquivar']",
        desarquivarTopoX: "//div[starts-with(@id, 'tbar-lista-mensagens')]//a[@data-qtip='Desarquivar']",
        marcarLidoTopoX: "//div[starts-with(@id, 'tbar-lista-mensagens')]//a[@data-qtip='Marcar como lido/não lido']",
        checkMarcarTodosX: "//div[starts-with(@id, 'tbar-lista-mensagens')]//a/span/span/span[starts-with(@id, 'use-smallsplitbutton')][1]",
        arrowCheckMarcarTodosX: "//div[starts-with(@id, 'tbar-lista-mensagens')]//a/span[starts-with(@id, 'use-smallsplitbutton')][2]",
        menuMarcarTodosX: "//div[starts-with(@id, 'menu')]//span[text()= 'Marcar todos']",
        menuDesmarcarTodosX: "//div[starts-with(@id, 'menu')]//span[text()= 'Desmarcar todos']",
        menuMarcarLidosX: "//div[starts-with(@id, 'menu')]//span[text()= 'Marcar lidos']",
        menuMarcarNaoLidosX: "//div[starts-with(@id, 'menu')]//span[text()= 'Marcar não lidos']",
        voltarTopoMsgX: "//div[starts-with(@id, 'panelleitura-geral-mensagens')][not(contains(@id, 'spam'))]//a[@data-qtip='Voltar']",
        arquivarTopoMsgX: "//div[starts-with(@id, 'panelleitura-geral-mensagens')][not(contains(@id, 'spam'))]//a[@data-qtip='Arquivar']",
        desarquivarTopoMsgX: "//div[starts-with(@id, 'panelleitura-geral-mensagens')][not(contains(@id, 'spam'))]//a[@data-qtip='Desarquivar']",
        marcarLidoTopoMsgX: "//div[starts-with(@id, 'panelleitura-geral-mensagens')][not(contains(@id, 'spam'))]//a[@data-qtip='Marcar como lido/não lido']",
        responderTopoMsgX: "//div[starts-with(@id, 'panelleitura-geral-mensagens')][not(contains(@id, 'spam'))]//a[@data-qtip='Responder']",
        responderATodosTopoMsgX: "//div[starts-with(@id, 'panelleitura-geral-mensagens')][not(contains(@id, 'spam'))]//a[@data-qtip='Responder a todos']",
        encaminharTopoMsgX: "//div[starts-with(@id, 'panelleitura-geral-mensagens')][not(contains(@id, 'spam'))]//a[@data-qtip='Encaminhar']",
        assuntoMsgX: "//div[starts-with(@id, 'panel-mensagens')]//div[starts-with(@id, 'panelleitura-geral-mensagens')]//label[contains(@style, 'display: block;')]",
        cabecalhoMsgX: "//div[@class='header-msg']",
        colFotoMsgX: "//div[@class='header-msg']/div[1]",
        colDeMsgX: "//div[@class='header-msg']/div[2]/div[1]",
        colParaMsgX: "//div[@class='header-msg']/div[2]/div[3]",
        colDataHoraMsgX: "//div[@class='header-msg']/div[3]",
        corpoMsgX: "//div[@class='content-msg']",
        responderMsgX: "//div[starts-with(@id, 'panel-mensagens')]//span[text()= 'Responder']",
        responderATodosMsgX: "//div[starts-with(@id, 'panel-mensagens')]//span[text()= 'Responder a todos']",
        encaminharMsgX: "//div[starts-with(@id, 'panel-mensagens')]//span[text()= 'Encaminhar']",
    },
}
